src/dctest/core.cljs

```diff
@@ -177,87 +179,119 @@ Options:
-                                        (interpolate command)
-                                        (mapv interpolate command)))))
-
-              run-attempt (fn []
-                            (P/let [results (run-exec step-context step)
-                                    step-context (assoc step-context :step results)
-                                    error (or (:error results)
-                                              (first (keep #(get-expect-error step-context %) (:expect step))))]
-                              (merge results
-                                     (when error {:error error}))))
-              {:keys [interval retries]} (:repeat step)
-              results (retry/retry-times run-attempt
-                                         retries
-                                         {:delay-ms interval
-                                          :check-fn #(not (:error %))})
-
-              outcome (if (:error results) :fail :pass)
-              results (merge results
-                             {:outcome outcome}
-                             (select-keys step [:name]))
-              context (update-in context [:state :failed] #(or % (failure? results)))]
+(defn get-expect-error [context expect]
+  (if (string? expect)
+    (when-not (expr/read-eval context expect)
+      {:message (str "Expectation failure: " expect)
+       :debug (expr/explain-refs context expect)})
+    (first (keep #(get-expect-error context %) expect))))
+
+(defn execute-step-attempts [{:keys [context results]}]
+  (P/let [{:keys [expect repeat]} results
+          {:keys [interval retries]} repeat
+          run-attempt (fn []
+                        (P/let [exec-results (run-exec context results)
+                                context (assoc context :step exec-results)
+                                error (or (:error exec-results)
+                                          (get-expect-error context expect))
+                                results (if error
+                                          (assoc results :outcome :fail :error error)
+                                          results)]
+                          {:context context :results results}))]
+    (retry/retry-times run-attempt
+                       retries
+                       {:delay-ms interval
+                        :check-fn #(not (failure? (:results %)))})))
+
+(defn execute-step [orig-context step]
+  (P/-> {:context orig-context
+         :results (assoc step :outcome :pending)}
+
+        (while-> (comp pending? :results)
+
+                 (read-eval-key :if)
+                 (update :results (fn [step]
+                                    (if (:if step)
+                                      step
+                                      (assoc step :outcome :skip))))
+
+                 (interpolate-key :env)
+                 (as-> m
+                   (update-in m [:context :env] merge (get-in m [:results :env])))
```

Comment:
All expressions in suites/test/steps are evaluated using the current context.  For instance, if you see `${{ env.FOO }}` the `env` is coming from the context `:env` map.  So any time a suite/test/step has env settings, we need to resolve them (in this case using `interpolate-key`) and then merge the resolved values back into the context env.  Perhaps there's a clearer way to do this?

Reply:
From just looking at this function, I'm unable to tell whether it's important that the interpolated env be in both the step (result) and the context because it's put in one place and then copied to the other.  So is it needed in both?  To answer this question, I need to dig deeper.  I think it only needs to be in context.  

So could these two expressions be replaced by:

```clojure
(as-> m
   (update-in m [:context :env] (interpolate-key :env))
```

I know it can't, b/c the present `interpolate-key` potentially does more than just interpolate the key (e.g. in the case of error).  I guess the moral is 'monads are hard' -- especially when not all functions are built around them (e.g. `update-in`).

Reply:
This is somewhat related to the discussion in the other thread, where we have to "pop" env off of the context, when we're done.  I think it would actually be cleaner if the interpolation _only_ happened in the suite/test/step, and then when we go to read-eval/interpolate, we combine the env's from that stack at the last second.  There are some details/challenges there that seemed to make this path messy, but perhaps I should give that another shot.